File: packages/TextPrompts.jl/src/TextPrompts.jl
"""
    TextPrompts

A minimal, zero-complexity prompt loader with TOML/YAML front-matter metadata
and section parsing.

TextPrompts allows you to manage prompts as text files with optional metadata,
providing safe string formatting, section extraction, and flexible loading.

# Quick Start

```julia
using TextPrompts

# Load a prompt file
prompt = load_prompt("prompts/greeting.txt")

# Format with placeholders
result = prompt(; name="World", day="Monday")

# Load a specific section
section = load_section("prompts/multi.txt", "system")

# Parse sections from a document
result = parse_sections(read("doc.md", String))
```
"""
module TextPrompts

using Dates
using TOML
using YAML

# Export types
export MetadataMode, STRICT, ALLOW, IGNORE
export PromptMeta, PromptString, Prompt
export Section, Link, FrontmatterBlock, ParseResult

# Export functions
export from_path, from_string
export load_prompt, load_section, save_prompt
# Note: `format` is intentionally NOT exported to avoid clashes with Dates.format etc.
# Use the callable syntax instead: prompt(; name="value") or TextPrompts.format(...)
export set_metadata, get_metadata, skip_metadata, warn_on_ignored_metadata
export extract_placeholders, get_placeholder_info, validate_format_args
export parse_sections, generate_slug, normalize_anchor_id
export inject_anchors, render_toc, get_section_text

# Include source files in dependency order
include("errors.jl")
include("config.jl")
include("placeholders.jl")
include("types.jl")
include("parser.jl")
include("sections.jl")
include("loaders.jl")
include("savers.jl")

# Initialize from environment on module load
function __init__()
    __init_config__()
end

end # module TextPrompts

File: packages/textprompts-ex/lib/text_prompts.ex
defmodule TextPrompts do
  @moduledoc """
  Public facade for the TextPrompts Elixir port.
  """

  alias TextPrompts.{Loader, Prompt, Saver, Sections}

  @type option :: {:meta, TextPrompts.MetadataMode.t()}

  @spec load_prompt(Path.t(), keyword()) :: {:ok, Prompt.t()} | {:error, Exception.t()}
  defdelegate load_prompt(path, opts \\ []), to: Loader

  @spec load_prompt!(Path.t(), keyword()) :: Prompt.t()
  def load_prompt!(path, opts \\ []) do
    case load_prompt(path, opts) do
      {:ok, prompt} -> prompt
      {:error, error} -> raise error
    end
  end

  @spec save_prompt(Path.t(), Prompt.t() | String.t(), keyword()) :: :ok | {:error, Exception.t()}
  defdelegate save_prompt(path, prompt, opts \\ []), to: Saver

  @spec save_prompt!(Path.t(), Prompt.t() | String.t(), keyword()) :: :ok
  def save_prompt!(path, prompt, opts \\ []) do
    case save_prompt(path, prompt, opts) do
      :ok -> :ok
      {:error, error} -> raise error
    end
  end

  @spec parse_sections(String.t()) :: Sections.ParseResult.t()
  defdelegate parse_sections(text), to: Sections

  @spec generate_slug(String.t()) :: String.t()
  defdelegate generate_slug(text), to: Sections

  @spec normalize_anchor_id(String.t()) :: String.t()
  defdelegate normalize_anchor_id(text), to: Sections

  @spec get_section_text(String.t(), String.t()) :: {String.t() | nil, boolean()}
  defdelegate get_section_text(text, anchor), to: Sections
end
